rust/cyclotron-core/src/worker.rs

```diff
@@ -24,29 +26,91 @@ pub struct Worker {
-    pending: Arc<Mutex<HashMap<Uuid, JobUpdate>>>,
+    // rather than having to track the update state on their side and submit it all at once.
+    // This also lets us "hide" all the locking logic, which we're not totally settled on yet.
+
+    // TRICKY - this is a sync mutex, because that simplifies using the manager in an FFI
+    // context (since most functions below can be sync). We have to be careful never to
+    // hold a lock across an await point, though.
+    pending: Mutex<HashMap<Uuid, JobUpdate>>,
+    // Metadata includes things like shard_id, which is likely to be used in latency sensitive
+    // contexts like reporting metrics. For this reason, we cache it, and are also careful to make
+    // the "happy path" of cache hits fast. For this reason, we use tokio::RwLock, to let us hold
+    // the lock over DB fetches. This does make things a bit awkward for FFI callers.
+    metadata: RwLock<HashMap<String, MetadataCacheEntry>>,
+}
+
+struct MetadataCacheEntry {
+    pub value: Option<String>,
+    pub last_set: DateTime<Utc>,
+}
+
+impl MetadataCacheEntry {
+    pub fn new(value: Option<String>, last_set: DateTime<Utc>) -> Self {
+        Self { value, last_set }
+    }
+
+    pub fn is_expired(&self) -> bool {
+        (Utc::now() - self.last_set).num_seconds() > 60
+    }
 }
 
 impl Worker {
     pub async fn new(config: PoolConfig) -> Result<Self, QueueError> {
         let pool = config.connect().await?;
         Ok(Self {
             pool,
-            pending: Arc::new(Mutex::new(HashMap::new())),
+            pending: Default::default(),
+            metadata: Default::default(),
         })
     }
 
     pub fn from_pool(pool: PgPool) -> Self {
         Self {
             pool,
-            pending: Arc::new(Mutex::new(HashMap::new())),
+            pending: Default::default(),
+            metadata: Default::default(),
+        }
+    }
+
+    pub async fn shard_id(&self) -> Result<Option<String>, QueueError> {
+        let metadata = self.metadata.read().await;
+        // A None value here means we've never hit the DB. A Some(None) means we got no result.
+        let shard_id = match metadata.get(SHARD_ID_KEY) {
+            Some(entry) => {
+                if entry.is_expired() {
+                    drop(metadata);
+                    self.fetch_metadata(SHARD_ID_KEY).await?
+                } else {
+                    entry.value.clone()
+                }
+            }
+            None => {
+                drop(metadata);
+                self.fetch_metadata(SHARD_ID_KEY).await?
+            }
+        };
+        Ok(shard_id)
+    }
+
+    // Actually hit the DB to fetch some metadata
+    async fn fetch_metadata(&self, key: &str) -> Result<Option<String>, QueueError> {
+        let mut metadata = self.metadata.write().await;
+
+        // Multiple tasks can race to update metadata, so double check someone didn't beat us
+        // to it
```

Comment:
We also `await` below, which could introduce more races, but I guess the idea is it's safe for them all to `insert`?

Reply:
Not relevant now, but because we hold the write lock across the DB fetch await point, other tasks racing to fetch the metadata would have been stalled on line 98, so when the winning task updated the map, all tasks that subsequently acquired the write lock would then see the updated value and short-circuit rather than inserting